pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,523 @@
+    Examples for this are shift values or frequencies that lie very close to each other.
+
+    **Examples**
+
+    Consider the following QNode:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x, Y, f=1.0):
+            qml.RX(f*x, wires=0)
+            qml.RY(Y[0], wires=0)
+            qml.RY(Y[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RY(5*Y[1], wires=1)
+            return qml.expval(qml.PauliZ(0)@qml.PauliZ(1))
+
+        x = 0.4
+        Y = np.array([1.9, -0.5])
+        f = 2.3
+
+        circuit_value = circuit(x, Y)
+
+    It has three variational parameters ``x`` (a scalar) and two entries of ``Y``
+    (an array-like) as well as a tunable frequency ``f`` for the ``qml.RX`` rotation.
+    A first reconstruction job could be with respect to the first entry of ``Y``,
+    which enters the circuit with a single integer frequency:
+
+    >>> nums_frequency = {"Y": {0: 1, 1: 6}}
+    >>> rec = qml.fourier.reconstruct(circuit, {"Y": [0]}, nums_frequency)(x, Y)
+    >>> rec
+    {'Y': {0: <function _reconstruct_equ.<locals>._reconstruction at 0x7f949627d790>}}
+    >>> recon_Y0 = rec["Y"][0]
+    >>> np.isclose(recon_Y0(Y[0]), circuit_value)
+    True
+    >>> np.isclose(recon_Y0(Y[0]+1.3), circuit(x, Y+np.eye(2)[0]*1.3)
+    True
+
+    We successfully reconstructed the dependence on ``Y[0]`` , keeping ``x`` and ``Y[1]``
+    at the values initialized above. Note that even though information about ``Y[1]`` is
+    contained in ``nums_frequency`` , ``ids`` determines which reconstructions are performed.
+    We may do the same for ``Y[1]`` , which enters the circuit with maximal frequency
+    :math:`1+5=4` . We will also track the number of executions needed:
+
+    >>> dev._num_executions = 0
+    >>> rec = qml.fourier.reconstruct(circuit, {"Y": [1]}, nums_frequency)(x, Y)
+    >>> dev.num_executions
+    13
+
+    As expected, we required :math:`2R+1=2\cdot 6+1=13` circuit executions. However, not
+    all frequencies below :math:`f_\text{max}=6` are present in the circuit, so that
+    a reconstruction using knowledge of the spectrum will be cheaper:
+
+    >>> dev._num_executions = 0
+    >>> spectra = {"Y": {1: [0., 1., 4., 5., 6.]}}
+    >>> rec = qml.fourier.reconstruct(circuit, {"Y": [1]}, None, spectra)(x, Y)
+    >>> dev.num_executions
+    9
+
+    If we want to reconstruct the dependence of ``circuit`` on ``x`` , we can not use
+    ``nums_frequency`` if ``f`` is not an integer. One could rescale ``x`` to obtain
+    the frequency :math:`1` again, or directly use ``spectra`` . We will combine this
+    with another reconstruction with respect to ``Y[0]`` :
+
+    >>> dev._num_executions = 0
+    >>> spectra = {"x": {0: [0., f]}, "Y": {0: [0., 1.]}}
+    >>> rec = qml.fourier.reconstruct(circuit, None, None, spectra)(x, Y, f=f)
+    >>> dev.num_executions
+    5
+    >>> recon_x = rec["x"][0]
```

Comment:
Ohhhh. It changes depending if you use `spectra` or `num_frequencies`.

Reply:
This I can't reproduce I think. In which scenario did `rec["x"][0]` not work but `rec["x"][()]` did?
re: indexing, this is the old story of why thought about introducing `ArgMap`. The doubled index is basically the
price we pay for keeping it a simple data structure. But anyways, for the inputs this is the same and even though it's
not the nicest API, it is okay, I believe. I'd be happy to see a more intuitive version that does not try to infer a lot about the arguments as `ArgMap` did :)

Reply:
It happens with this program here:
```python
dev = qml.device("default.qubit", wires=2)

@qml.beta.qnode(dev, diff_method="parameter-shift")
def circuit(x, Y):
    qml.RX(x, wires=0)
    qml.RY(Y[0], wires=0)
    qml.RY(Y[1], wires=1)
    qml.CNOT(wires=[0, 1])
    qml.RY(5 * Y[1], wires=1)
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

x = np.array(0.4, requires_grad=True)
y = np.array(1.6, requires_grad=True)

Y = np.array([1.9, -0.5], requires_grad=True)
spectrum = qml.fourier.qnode_spectrum(circuit, argnum=0)(x, Y)
recon_fns = qml.fourier.reconstruct(circuit, spectra=spectrum)(x, Y)

cost = recon_fns["x"][()]  # index using ()

loss = cost(x)
grad = qml.grad(cost)(x)
```

Reply:
But then, if I switch away from `spectra`, and instead specify `nums_frequency`, I have to index using `0`:

```python
freq = np.array(spectrum["x"][()])
num_freq = {"x": {0: len(freq[freq > 0])}}
recon_fns = qml.fourier.reconstruct(circuit, nums_frequency=num_freq)(x, Y)

cost = recon_fns["x"][0]  # index using 0

loss = cost(x)
grad = qml.grad(cost)(x)
```

Reply:
corollary: the former (using spectra) gives the correct loss and gradient. The latter (using `nums_frequency`) gives the correct loss, but `NaN` for the grad:

```
/home/josh/xanadu/pennylane/pennylane/numpy/tensor.py:155: RuntimeWarning: invalid value encountered in true_divide
  res = super().__array_ufunc__(ufunc, method, *args, **kwargs)
nan
```

Reply:
Ah, okay, I found the problem, thanks so much for catching this! Indeed I switched the default parameter index for scalar inputs in my head for some reason. It will be `()` as output by `qnode_spectrum`.